app/controllers/registrations_controller.rb

```diff
@@ -822,19 +823,69 @@ def create
   end
 
   def create_paypal_order
-    return head :forbidden if Rails.env.production?
+    return head :forbidden if PaypalInterface.paypal_disabled?
 
     @registration = registration_from_params
-    render json: PaypalInterface.create_order(@registration)
+    render json: PaypalInterface.create_order(@registration, params[:total_charge])
   end
 
   def capture_paypal_payment
-    return head :forbidden if Rails.env.production?
+    return head :forbidden if PaypalInterface.paypal_disabled?
 
     @registration = registration_from_params
     @competition = @registration.competition
     order_id = params[:order_id]
 
-    render json: PaypalInterface.capture_payment(@competition, order_id)
+    response = PaypalInterface.capture_payment(@competition, order_id)
+    if response['status'] == 'COMPLETED'
+
+      # TODO: Handle the case where there are multiple captures for a payment
+      # 1) Multiple installments
+      # 2) Some failed, some succeeded
+
+      amount_details = response['purchase_units'][0]['payments']['captures'][0]['amount']
+      currency_code = amount_details['currency_code']
+      amount = PaypalRecord.amount_in_cents(amount_details["value"], currency_code)
+      record = PaypalRecord.find_by(record_id: response["id"])
+
+      # Create a Capture object and link it to the PaypalRecord
+      # NOTE: This assumes there is only ONE capture per order - not a valid long-term assumption
+      capture_from_response = response['purchase_units'][0]['payments']['captures'][0]
+      PaypalRecord.create(
+        record_id: capture_from_response['id'],
+        status: capture_from_response['status'],
+        payload: {}, # TODO: Refactor so that we can actually capture the payload? Perhaps this needs to be called in PaypalInterface?
+        amount_in_cents: capture_from_response['amount']['value'],
+        currency_code: capture_from_response['amount']['currency_code'],
+        record_type: :capture,
+        parent_record: record,
+      )
+
+      # Record the payment
+      @registration.record_payment(
+        amount,
+        currency_code,
+        record, # TODO: Add error handling for the PaypalRecord not being found
+        @registration.user.id,
+      )
+    end
+
+    render json: response
+  end
+
+  def refund_paypal_payment
+    registration = Registration.find(params[:id])
+    paypal_order = RegistrationPayment.find(params[:payment_id]).receipt
```

Comment:
Nit: We could pass payment.receipt directly in the function that calls the refund endpoint - I tried that briefly and got errors so I went back to what I knew worked.

Not sure what best practice would be in terms of which object to pass - but the fact that we use the id we're given to immediately call another object suggests that I'm passing in the wrong object here

Reply:
No, it does not. It only suggests that the frontend is agnostic about the payment service being used. And that's a good thing.

There is nothing to worry about here, this is perfectly fine polymorphic modelling at play.